src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerContext.cs

```diff
@@ -37,9 +37,12 @@ internal bool CanUseSerializationLogic
                             // Guard against unsupported features
                             Options.Converters.Count == 0 &&
                             Options.Encoder == null &&
-                            Options.NumberHandling == JsonNumberHandling.Strict &&
+                            (((JsonNumberHandling.WriteAsString | JsonNumberHandling.AllowNamedFloatingPointLiterals) & Options.NumberHandling) == 0) &&
```

Comment:
If I understand the bug correctly, it is us miscalculating `_canUserSerializationLogic` in cases where we have `Options.NumberHandling == JsonNumberHandling.AllowReadingFromString`?

Also, nit:
```suggestion
                            ((Options.NumberHandling & (JsonNumberHandling.WriteAsString | JsonNumberHandling.AllowNamedFloatingPointLiterals)) == 0) &&
```

Reply:
Yes `AllowReadingFromString` has no bearing on serialization, so we can allow fast-path serialize logic to run when it is set.